### PR TITLE
perf(topn): reduce unnecessary table scan in streaming (Group)TopN executors

### LINKED ISSUE #13797: perf(streaming): optimzie topN cache
When the streaming TopN cache are deleted to no records, RW will try to scan from the table and fill the cache. But if there is no records in the table or group, TopN cache will scan from the table again and again. It is especially bad when the table has many delete tombstones.
The issue appeared in the groupTopN, the Agg's minput should have been solved with the row count.

@Little-Wallace found in NexMark q9 with temporal filter
image

### PR DESCRIPTION
**Pull request by stdrc**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

According to #13797 and some further investigation, when using a temporal filter with short time period, it's very likely that the `high` part of TopN cache frequently being deleted to empty. And because TopN executors (actually the _managed TopN state_) just blindly scan the state table to fill `high cache`, the performance can be significantly impacted.

This PR maintains a `row_count` in the _managed TopN state_, so that it's possible to avoid table scan when cache is empty if the managed state knowns the cache is still synced with the table (row counts match).

Fixes #13797.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/executor/top_n/top_n_state.rs [hunk 1]
```diff
@@ -53,16 +59,43 @@ impl<S: StateStore> ManagedTopNState<S> {
...
     pub fn insert(&mut self, value: impl Row) {
         self.state_table.insert(value);
+        if let Some(row_count) = self.row_count.as_mut() {
+            *row_count += 1;
+        }
     }
 
     pub fn delete(&mut self, value: impl Row) {
         self.state_table.delete(value);
+        if let Some(row_count) = self.row_count.as_mut() {
```
Comment (Little-Wallace) [inline src/stream/src/executor/top_n/top_n_state.rs:96]: Can we ensure that the delete key must exist in hummock ?

File: src/stream/src/executor/top_n/top_n_state.rs [hunk 2]
```diff
@@ -133,20 +172,29 @@ impl<S: StateStore> ManagedTopNState<S> {
...
             let topn_row = self.get_topn_row(item?.into_owned_row(), group_key.len());
             if let Some(start_key) = start_key.as_ref()
                 && &topn_row.cache_key <= start_key
             {
                 continue;
             }
-            // let row= &topn_row.row;
             cache.insert(topn_row.cache_key, (&topn_row.row).into());
             if cache.len() == cache_size_limit {
+                table_row_count = None; // cache becomes full, we cannot get precise table row count this time
```
[the remark this answers is not in the document]
Comment (stdrc) [reply src/stream/src/executor/top_n/top_n_state.rs:189]: Wait a minute, I'm still working on this. Not ready😁

File: src/stream/src/executor/top_n/top_n_cache.rs [hunk 3]
```diff
@@ -181,6 +187,21 @@ impl<const WITH_TIES: bool> TopNCache<WITH_TIES> {
...
+    /// Get total count of entries in the cache.
+    pub fn len(&self) -> usize {
+        self.low.len() + self.middle.len() + self.high.len()
+    }
+
+    pub(super) fn update_table_row_count(&mut self, table_row_count: usize) {
+        self.table_row_count = Some(table_row_count)
+    }
+
+    fn table_row_count_matched(&self) -> bool {
```
Comment (xxchan) [inline src/stream/src/executor/top_n/top_n_cache.rs:199]: May add some comments about what it implies when they match or mismatch

File: src/stream/src/executor/top_n/top_n_cache.rs [hunk 4]
```diff
@@ -325,19 +351,20 @@ impl TopNCacheTrait for TopNCache<false> {
             && (self.offset == 0 || cache_key > *self.low.last_key_value().unwrap().0)
         {
             // The row is in mid
+            self.middle.remove(&cache_key);
```
Comment (xxchan) [inline src/stream/src/executor/top_n/top_n_cache.rs:354]: why to move this?

File: src/stream/src/executor/top_n/top_n_state.rs [hunk 5]
```diff
@@ -114,13 +137,14 @@ impl<S: StateStore> ManagedTopNState<S> {
     /// * `start_key` - The start point of the key to scan. It should be the last key of the middle
     ///   cache. It doesn't contain the group key.
     pub async fn fill_high_cache<const WITH_TIES: bool>(
-        &self,
+        &mut self,
```
[the remark this answers is not in the document]
Comment (stdrc) [reply src/stream/src/executor/top_n/top_n_state.rs:140]: Yes, will remove

File: src/stream/src/executor/top_n/top_n_cache.rs (hunk 3 above)
Comment (stdrc) [reply src/stream/src/executor/top_n/top_n_cache.rs:199]: In `fill_high_cache` we scan the whole group in the table, skipping rows that below `high`, this PR counts these skipped row as well.

However e2e failed, I'm still debugging.

**Comment by xxchan** [comment]
Did you tested the perf issue is resolved by this?

**Comment by stdrc** [comment]
> Did you tested the perf issue is resolved by this?

Will run a test soon

File: src/stream/src/executor/top_n/top_n_state.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stdrc) [reply src/stream/src/executor/top_n/top_n_state.rs]: I don't quite get it. When `fill_high_cache` being called, the `high` cache will never be full, so this part of code is just to insert into the cache until it reaches the size limit.

**Comment by stdrc** [comment]
截屏2023-12-11 13 15 40

截屏2023-12-11 13 16 13

The throughput doesn't change too much while the bloom filter false-positive rate does decrease compared to #13797.

cc @st1page @Little-Wallace

**Comment by Little-Wallace** [comment]
> The throughput doesn't change too much while the bloom filter false-positive rate does decrease compared to https://github.com/risingwavelabs/risingwave/issues/13797.

Thanks to your work. Could you run a test again because https://github.com/risingwavelabs/risingwave/pull/13558/ have been merged and it may reduce some IO operations.

**Comment by stdrc** [comment]
> > The throughput doesn't change too much while the bloom filter false-positive rate does decrease compared to #13797.
> 
> Thanks to your work. Could you run a test again because #13558 have been merged and it may reduce some IO operations.

截屏2023-12-12 01 36 13
截屏2023-12-12 01 36 41

Something was going wrong...

**Comment by stdrc** [comment]
截屏2023-12-13 11 12 12

Now FPR drops to near zero.

**Comment by st1page** [comment]
> > > The throughput doesn't change too much while the bloom filter false-positive rate does decrease compared to #13797.
> > 
> > 
> > Thanks to your work. Could you run a test again because #13558 have been merged and it may reduce some IO operations.
> 
> https://buildkite.com/risingwave-test/nexmark-benchmark/builds/2649#018c597f-5684-4b84-a7db-eddfc753f995
> 
> Something was going wrong...
> 
> 截屏2023-12-12 01 36 13 截屏2023-12-12 01 36 41


The barrier's piling up is because at the beginning, the source emits insert operations, and after 5 minutes passed, the temporal filter began to emit delete operations. Then the performance degraded, source executors were backpressured. But the temporal filter can not be back pressured so it still emits the delete message.
Is the https://github.com/risingwavelabs/risingwave/pull/13271 included in your branch? If no, we can merge main and take a look.

**Comment by stdrc** [comment]
> Is the #13271 included in your branch? If no, we can merge main and take a look.

It is included🥵

**Comment by Little-Wallace** [comment]
Although the throughput does not increase much, this PR still reduce IOPS of hummock and FPR:
image

**Comment by st1page** [comment]
The low performance might because of https://github.com/risingwavelabs/risingwave/issues/13968

**Review by st1page — APPROVED** [review]
Let's merge the PR first because it actually solve a part of performance issue.

**Comment by fuyufjh** [comment]
Hey @st1page @stdrc Thanks for the work! Which query are you testing with? Wondering how much it solves the performance issue.

**Comment by stdrc** [comment]
> Which query are you testing with?

nexmark `q9-temporal-filter`, the main contribution of this PR was to reduce table iter ops

**Comment by fuyufjh** [comment]
> > Which query are you testing with?
> 
> nexmark `q9-temporal-filter`, the main contribution of this PR was to reduce table iter ops

Can you share the effect? like how much table iter ops was reduced after this PR

**Comment by stdrc** [comment]
> > > Which query are you testing with?
> > 
> > 
> > nexmark `q9-temporal-filter`, the main contribution of this PR was to reduce table iter ops
> 
> Can you share the effect? like how much table iter ops was reduced after this PR

As originally reported in https://github.com/risingwavelabs/risingwave/issues/13797, the bloom filter false-positive rate is high (up to 50% in most time), which means near half of table iter operations are wasted.

After this PR, the FPR is near 0% in most time, around 20% when warming up (many groups are new, so that manually maintained in-executor `table_row_count`s don't help). So we think that this PR did solve an aspect of the issue.

However, throughput was still not improved as expected. As @st1page found, another problem (possibly the true IO bottleneck) is that GroupTopN doesn't fetch missing groups concurrently (https://github.com/risingwavelabs/risingwave/issues/13968), which seems likely to prevent from storage layer batching.

**Pull request merged**